File: graph_2.cpp
/*
Problem #2: Adjacency Hashset Representation
● Can we iterate on neighbours efficiently (O(Degree )), while still checking whether an edge exists or not in O(1)?
  The best of 2 worlds? 
● Simply, yes. Instead of a list per node, use a hashset! 
● typedef vector<unordered_set<int>> GRAPH; 
   ○ Assume it is unweighted graph 
- Implement : void add_directed_edge(...), void add_undirected_edge, void print(...)
*/
#include<bits/stdc++.h>
using namespace std;

typedef vector<unordered_set<int>> GRAPH;

void add_directed_edge(GRAPH &graph, int from, int to)
{
    graph[from].insert(to);
}

void add_undirected_edge(GRAPH &graph, int from, int to)
{
    graph[from].insert(to);
    graph[to].insert(from);
}

void print(GRAPH &graph)
{
    int i = 0;
    for(unordered_set<int> &u: graph)
    {
        cout << i++ << " has " << u.size() << " neighbours : ";
        for(const int &v : u)
            cout << v << " ";
        cout << endl;
    }
}

int main()
{
    int nodes, edges, from, to;
    cin >> nodes >> edges;

    GRAPH graph(nodes);
    for(int i=0; i<edges; i++)
    {
        cin >> from >> to;
        add_directed_edge(graph, from, to);
    }
    cout << endl;
    print(graph); 
    return 0;
}

/*

6 7
2 1
2 5
2 0
2 3
0 5
1 4
5 4

Output
Node 0 has 1 neighbors: 5
Node 1 has 1 neighbors: 4
Node 2 has 4 neighbors: 3 0 1 5
Node 3 has 0 neighbors:
Node 4 has 0 neighbors:
Node 5 has 1 neighbors: 4

Space Complexity is O(E) E = Total Edges
Time complexity
	O(1) for add/remove/check edge
	O(neighbours) for iterating on neighbors

Disadvantages
	- No order guarantee for edges
	- Impractical for multiple edges  (we may do workarounds)
	- Hash tables extra memory
	- Like any hash tables, more caution to properly deal with the load (load factor)

 */

File: graph_2.py
''' 
Problem #2: Adjacency Hashset Representation             USING   defaultdict(set) 
● Can we iterate on neighbours efficiently (O(Degree )), while still checking whether an edge exists or not in O(1)?
  The best of 2 worlds? 
● Simply, yes. Instead of a list per node, use a hashset! 
● typedef vector<unordered_set<int>> GRAPH; 
   ○ Assume it is unweighted graph 
- Implement : void add_directed_edge(...), void add_undirected_edge, void print(...)
'''
from collections import defaultdict

def add_directed_edge(graph:list, fromm:int, to:int) :
    graph[fromm].add(to)

def add_undirected_edge(graph:list, fromm:int, to:int) :
    graph[fromm].add(to)
    graph[to].add(fromm)

def print_adjacency_list(graph:list) :
    for node, neighbours in graph.items() :
        print(f"Node {node} has {len(neighbours)} neighbours : {', '.join(map(str, neighbours))}")

def Main() :
    graph = defaultdict(set)
    edges = int(input()) # since it's a dict, we don't need nodes to define the maximum node in our dict. For list
                         # it's needed to access any index. 
    for _ in range(edges):
        fromm, to = map(int, input().split())
        add_directed_edge(graph, fromm, to)
    
    print_adjacency_list(graph)

if __name__ == '__main__' :
    Main()

'''
Output :
5
0 1
2 5
3 8
1 4
2 7
Node 0 has 1 neighbours : 1
Node 2 has 2 neighbours : 5, 7
Node 3 has 1 neighbours : 8
Node 1 has 1 neighbours : 4
'''